### DISCUSSION TITLE
Is custom `value.unit` impossible outside of `base` module?

### DISCUSSION THREAD

**Discussion by n-gt** [apple/pkl#870] [Q&A] [answered]
I was reading the [Number](https://github.com/apple/pkl/blob/d9c65d484a95316736712f0e0db91c45b7fe8a78/stdlib/base.pkl#L567) definition and seems like all the units like `Duration` and `DataSize` are intertwined in there.

I used the same logic from there, excluding the `external` keyword:
```
class Size {
 value: Number
 unit: "mm"|"cm"|"inches"
}
```

but the magic of `10.mm` is not accessible to me. 

I must admit if this is really the case and only certain units defined by the language maintainers are to be then it's a really disappointing design decision to see. Units should be completely abstracted and kept separate or not exist at all.

**Comment by n-gt** [comment]
This limitation obviously has big implications for the language down the road, since it locks comparison and arithmetic operators to those specific base types.

somehow we would have things like `30.min + 2.s` with no standard implementation and won't be able to do `20.kg + 10.g` or `100.mm + 2.km` or `1.kg == 1000.g` etc? please tell me I missed a point and it's totally possible

**Comment by bioball** [comment, marked as answer]
`DataSize` and `Duration` are built into the language. Properties like `5.min` and `5.mb` indeed are built directly into `Number`, and it's not possible to create your own properties that can be accessed off of `Number`.

Feel free to file an issue for a Size type (probably more accurately called `Length`) to be added to the language. The workaround for now, though, is to define your own class like you did here.

Note: being able to define your own types that are used like this would require [extensions](https://en.wikipedia.org/wiki/Extension_method) as well as [operator overloading](https://en.wikipedia.org/wiki/Operator_overloading). Generally, we are hesitant to add new language features for the fear of making the language too complicated; Pkl should be easy to learn and use. Also, extension methods are challenging for interpreted languages because variable resolution happens at runtime (you pay the cost of the lookups every time you eval the program, versus once during compilation).
